llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -824,7 +824,13 @@ namespace llvm {
 Value *createStepForVF(IRBuilderBase &B, Type *Ty, ElementCount VF,
                        int64_t Step) {
   assert(Ty->isIntegerTy() && "Expected an integer step");
-  return B.CreateElementCount(Ty, VF.multiplyCoefficientBy(Step));
+  if (!VF.isScalable() || !isPowerOf2_64(VF.getKnownMinValue()) ||
+      !isPowerOf2_64(Step))
+    return B.CreateElementCount(Ty, VF.multiplyCoefficientBy(Step));
+
+  return B.CreateShl(
+      B.CreateVScale(Ty),
+      ConstantInt::get(Ty, Log2_64((VF * Step).getKnownMinValue())), "", true);
```

Comment:
Minor nit, would it be slightly easier to read to early return in the optimised case, i.e.
```suggestion
  if (VF.isScalable() && isPowerOf2_64(VF.getKnownMinValue()) &&
      isPowerOf2_64(Step))
    return B.CreateShl(
        B.CreateVScale(Ty),
        ConstantInt::get(Ty, Log2_64((VF * Step).getKnownMinValue())), "", true);

  return B.CreateElementCount(Ty, VF.multiplyCoefficientBy(Step));
```

Not opinionated about this, up to you